### ISSUE TITLE
Invalid xpubs in PSBT

### ISSUE THREAD

**Issue by JamieDriver** [nunchuk-io/libnunchuk#20]
During testing Nunchuk with Blockstream Jade using QR codes and PSBTs, we encountered errors with the xpubs in the PSBTs.

For example - we scanned a QR which decoded to this PSBT (which Jade rejects as invalid) :  
```
"cHNidP8BAIcCAAAAAiDiCm67KgG3/xow836rt4/nKAY7Bstv4wV/H/2wC8UEAQAAAAD9////7D9IF6//Hcm+8BbqTG9jFeeIwnL9xoi1EOzwY4+83QQDAAAAAP3///8B6AMAAAAAAAAiACBJXNIf1pJyW8MPQKXGF1+M61RwrMa1RtxoAKNeM5/ydAAAAABPAQSIsh4DJJGS0oAAAAJmQ7vwDF3qgbPJaCqWauyzu2vFIEn3KJgyLwb8jVgRTwJVDyAKyzHsEc8iHLGBy1s3kErwJOJLYQQZda/DDbInHRQkkZLSMAAAgAAAAIAAAACAAgAAgE8BBIiyHgNn+Q/8gAAAAvKPSjp5rHv5tRQnpKph69TMCYfAxHVoFXcCygR3EUziA8+k7ZOpeM7nIZZoM953XpiJL5AtPVu0PjLyUcugs4MQFGf5D/wwAACAAAAAgAAAAIACAACATwEEiLIeA7I3/p2AAAACjRANNOu4pgzTbtBk8JwsDtFpRaSs+5yBaWZUhVErAD4DBsLRDaSnxBMDAPpnPLIjfz/JRWZ/PHlUcVSQ2IMeUS0Usjf+nTAAAIAAAACAAAAAgAIAAIAAAQCJAgAAAAFJpVe6K2rqa+G8WfU9jqcX+/fc9KSDcbvLXvtNYUv0hAEAAAAA/f///wLoAwAAAAAAACIAIMScPD2JGs6qsNgIbMAwSs+QJNTGQIFLjtXkP8rvY9xrsQMAAAAAAAAiACDakvdmSy/SOjINKNP2JmqkNVkeGyj1pSF6K8+jNKRq5XBNBwABASuxAwAAAAAAACIAINqS92ZLL9I6Mg0o0/YmaqQ1WR4bKPWlIXorz6M0pGrlAQVpUiECjwfYcUlBFD53h8BXNQlkukiJp3g+JxY1n2Ydk/XwXqghAt7GaBCofiFoGZNu4uD7kdfNOCF6bvAUO7oJ1UNI2lfiIQPUcpd1z4chtnyKsA08zEvQ+Rm1TExqxeaOOe9kuDvV51OuIgYCjwfYcUlBFD53h8BXNQlkukiJp3g+JxY1n2Ydk/XwXqgcJJGS0jAAAIAAAACAAAAAgAIAAIABAAAAAgAAACIGAt7GaBCofiFoGZNu4uD7kdfNOCF6bvAUO7oJ1UNI2lfiHGf5D/wwAACAAAAAgAAAAIACAACAAQAAAAIAAAAiBgPUcpd1z4chtnyKsA08zEvQ+Rm1TExqxeaOOe9kuDvV5xyyN/6dMAAAgAAAAIAAAACAAgAAgAEAAAACAAAAAAEA/aIBAgAAAAWDjIPFz7Dgkuf/e13Xx/V2Sz+09/CAL0VbYtFYdnJs3AAAAAAA/f///5G9NIMwWKc31sfUCjLZGxW3nw+Wt6ZOLk0Yg+6OsRToAQAAAAD9////kb00gzBYpzfWx9QKMtkbFbefD5a3pk4uTRiD7o6xFOgEAAAAAP3///+RvTSDMFinN9bH1Aoy2RsVt58PlremTi5NGIPujrEU6AAAAAAA/f///97mWtzk5CZPuh+AIjDJOXLP++0Lidecn/zoZqoAkQQbAAAAAAD9////BbwCAAAAAAAAIgAgSVzSH9aSclvDD0ClxhdfjOtUcKzGtUbcaACjXjOf8nS8AgAAAAAAACIAIKzLX9UXNARCeRVdWNwkIb2bzvxlJZSkPwSxiht3EAmevAIAAAAAAAAiACCNjBML/8XyotwYHCkUH280tQGxTczgqgaotV6BEzkjk7wCAAAAAAAAIgAgCylaGEc0q016qWFr4h4f3pj3sQUN1XP0XstyFciNBQ4HAgAAAAAAABYAFDcpN++hh7oZ7pYhMqT887gMoSibyp4LAAEBK7wCAAAAAAAAIgAgCylaGEc0q016qWFr4h4f3pj3sQUN1XP0XstyFciNBQ4BBWlSIQLdH9SBE3NZbC8TIVkRd1hmPXdpkugiUnkbVdO347stqCEDNiGRYK+n5j2ivr3D1E+pKpjT3xz65wrDQYs5EwK6E3YhA2TeamnGhtWXb69Tg4iE7bCCjKfcFNV82G1dfmUj0AyWU64iBgLdH9SBE3NZbC8TIVkRd1hmPXdpkugiUnkbVdO347stqByyN/6dMAAAgAAAAIAAAACAAgAAgAAAAAAHAAAAIgYDNiGRYK+n5j2ivr3D1E+pKpjT3xz65wrDQYs5EwK6E3YcJJGS0jAAAIAAAACAAAAAgAIAAIAAAAAABwAAACIGA2TeamnGhtWXb69Tg4iE7bCCjKfcFNV82G1dfmUj0AyWHGf5D/wwAACAAAAAgAAAAIACAACAAAAAAAcAAAAAAQFpUiECCMzBww7SyajOpTu7t4zaEMHCGGYlW4IGtRpvyry/UPQhAorjOHIYOV4ODr/Avy1CDkSzvvj4us6rjSI5wBAoRruRIQOdG4M/Opo8DDRnDPgLvFz4xLZJmJJBGUN1pYbcAF21RlOuIgICCMzBww7SyajOpTu7t4zaEMHCGGYlW4IGtRpvyry/UPQcZ/kP/DAAAIAAAACAAAAAgAIAAIAAAAAABAAAACICAorjOHIYOV4ODr/Avy1CDkSzvvj4us6rjSI5wBAoRruRHLI3/p0wAACAAAAAgAAAAIACAACAAAAAAAQAAAAiAgOdG4M/Opo8DDRnDPgLvFz4xLZJmJJBGUN1pYbcAF21RhwkkZLSMAAAgAAAAIAAAACAAgAAgAAAAAAEAAAAAA=="
```

The node 'deserialisepsbt' yields:
```
{
  "tx": {
    "txid": "bc35b8e60bd0d99445fa6be7d71fe5c717d252dcb4c36d79b69992ec339ad952",
    "hash": "bc35b8e60bd0d99445fa6be7d71fe5c717d252dcb4c36d79b69992ec339ad952",
    "version": 2,
    "size": 135,
    "vsize": 135,
    "weight": 540,
    "locktime": 0,
    "vin": [
      {
        "txid": "04c50bb0fd1f7f05e36fcb063b0628e78fb7ab7ef3301affb7012abb6e0ae220",
        "vout": 1,
        "scriptSig": {
          "asm": "",
          "hex": ""
        },
        "sequence": 4294967293
      },
      {
        "txid": "04ddbc8f63f0ec10b588c6fd72c288e715636f4cea16f0bec91dffaf17483fec",
        "vout": 3,
        "scriptSig": {
          "asm": "",
          "hex": ""
        },
        "sequence": 4294967293
      }
    ],
    "vout": [
      {
        "value": 0.00001000,
        "n": 0,
        "scriptPubKey": {
          "asm": "0 495cd21fd692725bc30f40a5c6175f8ceb5470acc6b546dc6800a35e339ff274",
          "hex": "0020495cd21fd692725bc30f40a5c6175f8ceb5470acc6b546dc6800a35e339ff274",
          "address": "bcrt1qf9wdy87kjfe9hsc0gzjuv96l3n44gu9vc665dhrgqz34uvul7f6q5nnfqp",
          "type": "witness_v0_scripthash"
        }
      }
    ]
  },
  "unknown": {
    "010488b21e03249192d2800000026643bbf00c5dea81b3c9682a966aecb3bb6bc52049f72898322f06fc8d58114f02550f200acb31ec11cf221cb181cb5b37904af024e24b61041975afc30db2271d": "249192d230000080000000800000008002000080",
    "010488b21e0367f90ffc80000002f28f4a3a79ac7bf9b51427a4aa61ebd4cc0987c0c47568157702ca0477114ce203cfa4ed93a978cee721966833de775e98892f902d3d5bb43e32f251cba0b38310": "67f90ffc30000080000000800000008002000080",
    "010488b21e03b237fe9d800000028d100d34ebb8a60cd36ed064f09c2c0ed16945a4acfb9c8169665485512b003e0306c2d10da4a7c4130300fa673cb2237f3fc945667f3c7954715490d8831e512d": "b237fe9d30000080000000800000008002000080"
  },
  "inputs": [
    {
      "witness_utxo": {
        "amount": 0.00000945,
        "scriptPubKey": {
          "asm": "0 da92f7664b2fd23a320d28d3f6266aa435591e1b28f5a5217a2bcfa334a46ae5",
          "hex": "0020da92f7664b2fd23a320d28d3f6266aa435591e1b28f5a5217a2bcfa334a46ae5",
          "type": "witness_v0_scripthash",
          "address": "bcrt1qm2f0wejt9lfr5vsd9rflvfn25s64j8sm9r662gt69086xd9ydtjs4egqqj"
        }
      },
      "non_witness_utxo": {
        "txid": "04c50bb0fd1f7f05e36fcb063b0628e78fb7ab7ef3301affb7012abb6e0ae220",
        "hash": "04c50bb0fd1f7f05e36fcb063b0628e78fb7ab7ef3301affb7012abb6e0ae220",
        "version": 2,
        "size": 137,
        "vsize": 137,
        "weight": 548,
        "locktime": 478576,
        "vin": [
          {
            "txid": "84f44b614dfb5ecbbb7183a4f4dcf7fb17a78e3df559bce16bea6a2bba57a549",
            "vout": 1,
            "scriptSig": {
              "asm": "",
              "hex": ""
            },
            "sequence": 4294967293
          }
        ],
        "vout": [
          {
            "value": 0.00001000,
            "n": 0,
            "scriptPubKey": {
              "asm": "0 c49c3c3d891aceaab0d8086cc0304acf9024d4c640814b8ed5e43fcaef63dc6b",
              "hex": "0020c49c3c3d891aceaab0d8086cc0304acf9024d4c640814b8ed5e43fcaef63dc6b",
              "address": "bcrt1qcjwrc0vfrt824vxcppkvqvz2e7gzf4xxgzq5hrk4uslu4mmrm34s425383",
              "type": "witness_v0_scripthash"
            }
          },
          {
            "value": 0.00000945,
            "n": 1,
            "scriptPubKey": {
              "asm": "0 da92f7664b2fd23a320d28d3f6266aa435591e1b28f5a5217a2bcfa334a46ae5",
              "hex": "0020da92f7664b2fd23a320d28d3f6266aa435591e1b28f5a5217a2bcfa334a46ae5",
              "address": "bcrt1qm2f0wejt9lfr5vsd9rflvfn25s64j8sm9r662gt69086xd9ydtjs4egqqj",
              "type": "witness_v0_scripthash"
            }
          }
        ]
      },
      "witness_script": {
        "asm": "2 028f07d8714941143e7787c057350964ba4889a7783e2716359f661d93f5f05ea8 02dec66810a87e216819936ee2e0fb91d7cd38217a6ef0143bba09d54348da57e2 03d4729775cf8721b67c8ab00d3ccc4bd0f919b54c4c6ac5e68e39ef64b83bd5e7 3 OP_CHECKMULTISIG",
        "hex": "5221028f07d8714941143e7787c057350964ba4889a7783e2716359f661d93f5f05ea82102dec66810a87e216819936ee2e0fb91d7cd38217a6ef0143bba09d54348da57e22103d4729775cf8721b67c8ab00d3ccc4bd0f919b54c4c6ac5e68e39ef64b83bd5e753ae",
        "type": "multisig"
      },
      "bip32_derivs": [
        {
          "pubkey": "028f07d8714941143e7787c057350964ba4889a7783e2716359f661d93f5f05ea8",
          "master_fingerprint": "249192d2",
          "path": "m/48'/0'/0'/2'/1/2"
        },
        {
          "pubkey": "02dec66810a87e216819936ee2e0fb91d7cd38217a6ef0143bba09d54348da57e2",
          "master_fingerprint": "67f90ffc",
          "path": "m/48'/0'/0'/2'/1/2"
        },
        {
          "pubkey": "03d4729775cf8721b67c8ab00d3ccc4bd0f919b54c4c6ac5e68e39ef64b83bd5e7",
          "master_fingerprint": "b237fe9d",
          "path": "m/48'/0'/0'/2'/1/2"
        }
      ]
    },
    {
      "witness_utxo": {
        "amount": 0.00000700,
        "scriptPubKey": {
          "asm": "0 0b295a184734ab4d7aa9616be21e1fde98f7b1050dd573f45ecb7215c88d050e",
          "hex": "00200b295a184734ab4d7aa9616be21e1fde98f7b1050dd573f45ecb7215c88d050e",
          "type": "witness_v0_scripthash",
          "address": "bcrt1qpv545xz8xj45674fv947y8slm6v00vg9ph2h8az7edeptjydq58q2auwp2"
        }
      },
      "non_witness_utxo": {
        "txid": "04ddbc8f63f0ec10b588c6fd72c288e715636f4cea16f0bec91dffaf17483fec",
        "hash": "04ddbc8f63f0ec10b588c6fd72c288e715636f4cea16f0bec91dffaf17483fec",
        "version": 2,
        "size": 418,
        "vsize": 418,
        "weight": 1672,
        "locktime": 761546,
        "vin": [
          {
            "txid": "dc6c727658d1625b452f80f0f7b43f4b76f5c7d75d7bffe792e0b0cfc5838c83",
            "vout": 0,
            "scriptSig": {
              "asm": "",
              "hex": ""
            },
            "sequence": 4294967293
          },
          {
            "txid": "e814b18eee83184d2e4ea6b7960f9fb7151bd9320ad4c7d637a758308334bd91",
            "vout": 1,
            "scriptSig": {
              "asm": "",
              "hex": ""
            },
            "sequence": 4294967293
          },
          {
            "txid": "e814b18eee83184d2e4ea6b7960f9fb7151bd9320ad4c7d637a758308334bd91",
            "vout": 4,
            "scriptSig": {
              "asm": "",
              "hex": ""
            },
            "sequence": 4294967293
          },
          {
            "txid": "e814b18eee83184d2e4ea6b7960f9fb7151bd9320ad4c7d637a758308334bd91",
            "vout": 0,
            "scriptSig": {
              "asm": "",
              "hex": ""
            },
            "sequence": 4294967293
          },
          {
            "txid": "1b049100aa66e8fc9f9cd7890bedfbcf7239c93022801fba4f26e4e4dc5ae6de",
            "vout": 0,
            "scriptSig": {
              "asm": "",
              "hex": ""
            },
            "sequence": 4294967293
          }
        ],
        "vout": [
          {
            "value": 0.00000700,
            "n": 0,
            "scriptPubKey": {
              "asm": "0 495cd21fd692725bc30f40a5c6175f8ceb5470acc6b546dc6800a35e339ff274",
              "hex": "0020495cd21fd692725bc30f40a5c6175f8ceb5470acc6b546dc6800a35e339ff274",
              "address": "bcrt1qf9wdy87kjfe9hsc0gzjuv96l3n44gu9vc665dhrgqz34uvul7f6q5nnfqp",
              "type": "witness_v0_scripthash"
            }
          },
          {
            "value": 0.00000700,
            "n": 1,
            "scriptPubKey": {
              "asm": "0 accb5fd51734044279155d58dc2421bd9bcefc652594a43f04b18a1b7710099e",
              "hex": "0020accb5fd51734044279155d58dc2421bd9bcefc652594a43f04b18a1b7710099e",
              "address": "bcrt1q4n94l4ghxszyy7g4t4vdcfpphkdualr9yk22g0cykx9pkacspx0qdkj7ny",
              "type": "witness_v0_scripthash"
            }
          },
          {
            "value": 0.00000700,
            "n": 2,
            "scriptPubKey": {
              "asm": "0 8d8c130bffc5f2a2dc181c29141f6f34b501b14dcce0aa06a8b55e8113392393",
              "hex": "00208d8c130bffc5f2a2dc181c29141f6f34b501b14dcce0aa06a8b55e8113392393",
              "address": "bcrt1q3kxpxzllche29hqcrs53g8m0xj6srv2dens25p4gk40gzyeeywfsfwjpap",
              "type": "witness_v0_scripthash"
            }
          },
          {
            "value": 0.00000700,
            "n": 3,
            "scriptPubKey": {
              "asm": "0 0b295a184734ab4d7aa9616be21e1fde98f7b1050dd573f45ecb7215c88d050e",
              "hex": "00200b295a184734ab4d7aa9616be21e1fde98f7b1050dd573f45ecb7215c88d050e",
              "address": "bcrt1qpv545xz8xj45674fv947y8slm6v00vg9ph2h8az7edeptjydq58q2auwp2",
              "type": "witness_v0_scripthash"
            }
          },
          {
            "value": 0.00000519,
            "n": 4,
            "scriptPubKey": {
              "asm": "0 372937efa187ba19ee962132a4fcf3b80ca1289b",
              "hex": "0014372937efa187ba19ee962132a4fcf3b80ca1289b",
              "address": "bcrt1qxu5n0maps7apnm5kyye2fl8nhqx2z2ymn4pc9x",
              "type": "witness_v0_keyhash"
            }
          }
        ]
      },
      "witness_script": {
        "asm": "2 02dd1fd4811373596c2f132159117758663d776992e82252791b55d3b7e3bb2da8 0336219160afa7e63da2bebdc3d44fa92a98d3df1cfae70ac3418b391302ba1376 0364de6a69c686d5976faf53838884edb0828ca7dc14d57cd86d5d7e6523d00c96 3 OP_CHECKMULTISIG",
        "hex": "522102dd1fd4811373596c2f132159117758663d776992e82252791b55d3b7e3bb2da8210336219160afa7e63da2bebdc3d44fa92a98d3df1cfae70ac3418b391302ba1376210364de6a69c686d5976faf53838884edb0828ca7dc14d57cd86d5d7e6523d00c9653ae",
        "type": "multisig"
      },
      "bip32_derivs": [
        {
          "pubkey": "02dd1fd4811373596c2f132159117758663d776992e82252791b55d3b7e3bb2da8",
          "master_fingerprint": "b237fe9d",
          "path": "m/48'/0'/0'/2'/0/7"
        },
        {
          "pubkey": "0336219160afa7e63da2bebdc3d44fa92a98d3df1cfae70ac3418b391302ba1376",
          "master_fingerprint": "249192d2",
          "path": "m/48'/0'/0'/2'/0/7"
        },
        {
          "pubkey": "0364de6a69c686d5976faf53838884edb0828ca7dc14d57cd86d5d7e6523d00c96",
          "master_fingerprint": "67f90ffc",
          "path": "m/48'/0'/0'/2'/0/7"
        }
      ]
    }
  ],
  "outputs": [
    {
      "witness_script": {
        "asm": "2 0208ccc1c30ed2c9a8cea53bbbb78cda10c1c21866255b8206b51a6fcabcbf50f4 028ae3387218395e0e0ebfc0bf2d420e44b3bef8f8baceab8d2239c0102846bb91 039d1b833f3a9a3c0c34670cf80bbc5cf8c4b649989241194375a586dc005db546 3 OP_CHECKMULTISIG",
        "hex": "52210208ccc1c30ed2c9a8cea53bbbb78cda10c1c21866255b8206b51a6fcabcbf50f421028ae3387218395e0e0ebfc0bf2d420e44b3bef8f8baceab8d2239c0102846bb9121039d1b833f3a9a3c0c34670cf80bbc5cf8c4b649989241194375a586dc005db54653ae",
        "type": "multisig"
      },
      "bip32_derivs": [
        {
          "pubkey": "0208ccc1c30ed2c9a8cea53bbbb78cda10c1c21866255b8206b51a6fcabcbf50f4",
          "master_fingerprint": "67f90ffc",
          "path": "m/48'/0'/0'/2'/0/4"
        },
        {
          "pubkey": "028ae3387218395e0e0ebfc0bf2d420e44b3bef8f8baceab8d2239c0102846bb91",
          "master_fingerprint": "b237fe9d",
          "path": "m/48'/0'/0'/2'/0/4"
        },
        {
          "pubkey": "039d1b833f3a9a3c0c34670cf80bbc5cf8c4b649989241194375a586dc005db546",
          "master_fingerprint": "249192d2",
          "path": "m/48'/0'/0'/2'/0/4"
        }
      ]
    }
  ],
  "fee": 0.00000645
}
```

The problem appears to be with the xpubs in the multisig quorum signers (listed as 'unknowns' in the above output).

eg (from the 'unknowns' above) :  
`"010488b21e03249192d2800000026643bbf00c5dea81b3c9682a966aecb3bb6bc52049f72898322f06fc8d58114f02550f200acb31ec11cf221cb181cb5b37904af024e24b61041975afc30db2271d": "249192d230000080000000800000008002000080",`

1. The first part (map key) should be the field flag `01` (xpub), then the serialised extended key.  
The serialised key starts with the type of key (`0488b21e03` = mainnet public key), and then the 'key depth' `03`, then the fingerprint `249192d2`, and so on.  

2. The map value part is the root master key fingerprint, `249192d2`, then the path data (8 hex digits per element): 0x `30000080/00000080/00000080/02000080` <- the fact there are 4 parts here implies the key depth is 4.

So the xpub (the map key) declares a depth of 3 - but the path (map value) shows a depth of 4 (due to the number of bytes).

So Jade (actually libwally-core) rejects this as invalid.

I note also that the xpub has the master key fingerprint in the 'parent hash' field - this is also not strictly correct, this should be the hash/fingperprint of the immediate parent of the given key (ie. 0x `30000080/00000080/00000080`)  - See the spec/test-vectors [here](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#Extended_keys)

**Comment by bakaoh** [comment]
Thank @JamieDriver , for serialised extended key, we decode it from xpub base58 using Bitcoin-core code so i guess there is no problem there. 

Jade xpub export using QR mode returns: 
```
[997a8fed/48'/0'/0'/2']xpub6CnCTux5YFbtA3DYTPZngdpAv7ctAP5KNLftUrTTK5w5TzwxkxoVTbYs1TQ63wtfVaecDbuw1yKLtnYnBaFnx9vQig18EYhpizApLqJQLtj
```

Using HWI with USB, it returns:
```
~/hwi -f 997a8fed  getxpub 'm/48h/0h/0h/2h'
{"xpub": "xpub6EPj5PJk8AL5fH46UgEzfp8tzFcHnNg9k5tdx6tYbgLYY5etuTeRCaxWfdPpX2zuysWtjR7pmLU6deWcfGuRdGY6N9BNVM5CbAbKkXgaPaq"}
```

The xpub returns by QR seems wrong, just as you said, its depth is 3 instead of 4, so if you using Jade at QR mode, it must be the cause

<img width="928" alt="image" src="https://user-images.githubusercontent.com/4920000/211524000-879be8c7-5a4b-4152-8487-1147614af3cc.png">

**Comment by JamieDriver** [comment]
Hi,
Jade QR export of xpub is in the BC 'ur:crypto-account' format rather than a straightforward base58 string - if there is an issue with that output/cbor that is something I can look into and you will have my humble apologies (and many thanks!).
We will investigate now and I'll get back to you asap.  Cheers.

**Comment by JamieDriver** [comment]
Many thanks for the analysis @bakaoh - I accept full responsibility!  ;-)  
It looks like we did indeed have some incorrectly-populated fields in the 'ur:crypto-account' which I see could lead you to produce those xpubs (which we would later reject as invalid).  
I have tweaked the QR export locally and can now work perfectly with Nunchuk+Jade signing multisig PSBTs.  
I will ensure this fix is rolled into our next firmware release.
Many thanks (and many apologies!)
